interp/iostream_test.go

```diff
@@ -0,0 +1,82 @@
+package interp
+
+import (
+	"errors"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"testing"
+)
+
+const (
+	outputStreamBufferSize = 1024
+)
+
+func TestStreamDoubleClose(t *testing.T) {
+	dir := t.TempDir()
+	t.Run("InFile", func(t *testing.T) {
+		f, err := os.Create(filepath.Join(dir, "infile"))
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer os.Remove(f.Name())
+		in := newInFileStream(f)
+		checkDoubleInputClose(t, in)
+	})
+	t.Run("OutFile", func(t *testing.T) {
+		f, err := os.Create(filepath.Join(dir, "outfile"))
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer os.Remove(f.Name())
+		out := newOutFileStream(f, outputStreamBufferSize)
+		checkDoubleOutputClose(t, out)
+	})
+	t.Run("InCmd", func(t *testing.T) {
+		cmd := execDefShell("echo close me")
+		in, err := newInCmdStream(cmd)
+		if err != nil {
+			t.Fatal(err)
+		}
+		checkDoubleInputClose(t, in)
+	})
+	t.Run("OutCmd", func(t *testing.T) {
+		cmd := execDefShell("echo close me")
+		out, err := newOutCmdStream(cmd)
+		if err != nil {
+			t.Fatal(err)
+		}
+		checkDoubleOutputClose(t, out)
+	})
+}
+
+func execDefShell(scriptlet string) *exec.Cmd {
+	cmdline := append(getDefaultShellCommand(), scriptlet)
+	return exec.Command(cmdline[0], cmdline[1:]...)
+}
+
+func checkDoubleInputClose(t *testing.T, in inputStream) {
+	if err := in.Close(); err != nil {
+		t.Fatal(err)
+	}
+	exitCode := in.ExitCode()
+	if err := in.Close(); !errors.Is(err, doubleCloseError) {
+		t.Error("expected inputStream.Close() to return error on double close")
+	}
+	if in.ExitCode() != exitCode {
+		t.Error("expected inputStream.ExitCode() to stay the same")
+	}
+}
+
+func checkDoubleOutputClose(t *testing.T, out outputStream) {
```

Comment:
We should probably mark these 2 check functions with `t.Helper()` so the stack frame / line number shows as expected in case of failure.

Reply:
Actually, it looks like these functions are identical apart from the type. Could add

```go
type inOutCloser {
    Close() error
    ExitCode() float64
}
```

And then combine them into the same function (both `inputStream` and `outputStream` implement that). Up to you.

Reply:
Ok.